File: test_ogex.py
#!/usr/bin/env python3
"""
Test file for Ogex Python bindings

First install the package:
  cd ogex-python
  maturin develop

Then run:
  python test_ogex.py
"""

import ogex

def test_basic_match():
    """Test basic pattern matching"""
    print("Test 1: Basic match")
    regex = ogex.compile("hello")
    print(f"  'hello world' matches: {regex.is_match('hello world')}")
    print(f"  'goodbye' matches: {regex.is_match('goodbye')}")

def test_named_groups():
    """Test named group capture"""
    print("\nTest 2: Named groups")
    regex = ogex.compile("(name:\\w+)")
    m = regex.search("hello name:John Smith")
    if m:
        print(f"  Match: '{m.text}'")
        print(f"  Named group 'name': '{m.group(1)}'")

def test_relative_backref():
    r"""Test relative backreference \g{-1}"""
    print("\nTest 3: Relative backreference \\g{-1}")
    regex = ogex.compile("(a)(b)\\g{-1}")
    print(f"  'abb' matches: {regex.is_match('abb')}")
    print(f"  'aba' matches: {regex.is_match('aba')}")

def test_g_literal():
    r"""Test \G as literal in pattern"""
    print("\nTest 4: \\G literal in pattern")
    regex = ogex.compile("\\G")
    print(f"  'G' matches: {regex.is_match('G')}")
    print(f"  'g' matches: {regex.is_match('g')}")

def test_findall():
    """Test findall functionality"""
    print("\nTest 5: Find all")
    regex = ogex.compile("\\d+")
    matches = regex.findall("abc 123 def 456 ghi 789")
    print(f"  Found {len(matches)} matches:")
    for i, m in enumerate(matches, 1):
        print(f"    {i}: '{m.text}' at {m.start}-{m.end}")

def test_sub():
    """Test substitution"""
    print("\nTest 6: Substitution")
    regex = ogex.compile("\\d+")
    result = regex.sub("[\\G]", "abc 123 def")
    print(f"  Replace digits with [match]: '{result}'")

if __name__ == "__main__":
    print("=== Ogex Python Tests ===\n")
    
    test_basic_match()
    test_named_groups()
    test_relative_backref()
    test_g_literal()
    test_findall()
    test_sub()
    
    print("\n✅ All tests completed!")

File: test_ogex.ts
/**
 * Test file for Ogex WASM bindings
 * 
 * Run with:
 *   node --experimental-strip-types test_ogex.ts
 *   deno run --allow-read --allow-net test_ogex.ts  
 *   bun test_ogex.ts
 */

// Import the WASM module
import init, { JsRegex } from "./ogex/pkg/ogex.js";

async function main() {
  // Initialize WASM
  await init();

  console.log("=== Ogex WASM Tests ===\n");

  // Test 1: Basic match
  console.log("Test 1: Basic match");
  const regex1 = new JsRegex("hello");
  console.log(`  "hello world" matches: ${regex1.is_match("hello world")}`);
  console.log(`  "goodbye" matches: ${regex1.is_match("goodbye")}`);

  // Test 2: Named groups
  console.log("\nTest 2: Named groups");
  const regex2 = new JsRegex("(name:\\w+)");
  const match2 = regex2.find("hello name:John Smith");
  if (match2) {
    console.log(`  Match: "${match2.text}"`);
    console.log(`  Named group 'name': "${match2.named_group("name")}"`);
  }

  // Test 3: Relative backreference \g{-1}
  console.log("\nTest 3: Relative backreference \\g{-1}");
  const regex3 = new JsRegex("(a)(b)\\g{-1}");
  console.log(`  "abb" matches: ${regex3.is_match("abb")}`);
  console.log(`  "aba" matches: ${regex3.is_match("aba")}`);

  // Test 4: \G literal in pattern
  console.log("\nTest 4: \\G literal in pattern");
  const regex4 = new JsRegex("\\G");
  console.log(`  "G" matches: ${regex4.is_match("G")}`);
  console.log(`  "g" matches: ${regex4.is_match("g")}`);

  // Test 5: Find all
  console.log("\nTest 5: Find all");
  const regex5 = new JsRegex("\\d+");
  const matches5 = regex5.find_all("abc 123 def 456 ghi 789");
  console.log(`  Found ${matches5.length} matches:`);
  for (let i = 0; i < matches5.length; i++) {
    const m = matches5[i];
    console.log(`    ${i + 1}: "${m.text}" at ${m.start}-${m.end}`);
  }

  // Test 6: Transpile
  console.log("\nTest 6: Transpile");
  const transpiled = JsRegex.transpile("(name:\\w+)(email:\\w+)");
  console.log(`  "(name:\\w+)(email:\\w+)" -> "${transpiled}"`);

  console.log("\n✅ All tests completed!");
}

main().catch(console.error);
